src/controllers/auth.controller.ts

```diff
@@ -0,0 +1,186 @@
+import admin from 'firebase-admin';
+import { Request, Response, NextFunction } from 'express';
+import { firebase as client } from '@firebase/app';
+import * as ERR_MSG from '../constants/ErrResponses';
+
+
+/*
+  Create a new auth object for the provided email and password.
+  Also links the auth account with existing user document and
+  updates document with provided profile information.
+  Prereqs:
+    1.  Request body contains the following key value pairs:
+        a. email
+        b. password
+        c. firstname
+        d. lastname
+        e. major
+        f. gradYear
+  Return: On successful account creation, user document is updated with incoming profile
+          information and user auth account object is sent back mapped to newUser.
+  Errors:`Can send back the following error responses: 400, 403, 500.
+*/
+export const signup = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<void> => {
+
+
+  if (!validInput(req.body)) {
+    // Invalid user information
+    res.status(400).json({
+      msg: ERR_MSG.GENERIC_400_MSG,
+    });
+    return next();
+  }
+
+  // Grab form information from request body
+  const { email,
+    password,
+    firstname,
+    lastname,
+    major,
+    gradYear } = req.body;
+
+  const userDocDetails: any = await getUserDocFromEmail(email);
+  const userDocID: string = userDocDetails.uid;
+
+  if (!userDocDetails) {
+    // email is not whitelisted.
+    res.status(403).json({
+      msg: ERR_MSG.USER_NOT_WHITELISTED,
+    });
+    return next();
+  }
+
+  let newUser: admin.auth.UserRecord;
+
+  try {
+    newUser = await admin.auth().createUser({
+      uid: userDocID,
+      email: email,
+      password: password
+    });
+
+    const updatedClaims: any = newUser.customClaims || {};
+    console.log("Initial User claims:");
+    console.log(newUser.customClaims);
+
+    // claims are either 'member', 'inductee' or 'officer
+    updatedClaims[userDocDetails.role.toLowerCase()] = true;
+    console.log("Updated User Claims:")
+    console.log(updatedClaims);
+    await admin.auth().setCustomUserClaims(userDocID, updatedClaims).catch(console.log);
+  }
+  catch (err) {
+    console.log(err);
+    if (err.code == 'auth/email-already-exists' || err.code == 'auth/uid-already-exists') {
+      // Account already exists.
+      res.status(403).json({
+        msg: ERR_MSG.USER_ALREADY_EXISTS,
+      });
+      console.log("account already exists");
+      return next();
+    }
+    else {
+      res.status(500).json({
+        msg: ERR_MSG.GENERIC_INTERNAL_ERROR,
+      });
+      console.log("Internal server error while creating auth");
+      return next();
+    }
+  }
+
+  // if (!newUser) {
+  //   return next();
+  // }
+
+  try {
+    await updateUserDoc('users', userDocID, {
+      first_name: firstname,
+      last_name: lastname,
+      major: major,
+      graduation_year: gradYear,
+    });
+
+    // Sign in to user account, send verification email and sign out.
```

Comment:
better to wrap these 3 lines in a single function to maintain abstraction for ease of change later when we migrate to email service

Reply:
Done in [this commit](https://github.com/HKN-UCSD/hkn-server/commit/4f7df3034bbcc36bcd5c58920eea8cd51eaf4fb2)